docs/proposals/component-registration.md

```diff
@@ -0,0 +1,409 @@
+    - Spec would have address (ip or host) and ports (optionally named) to connect with an external service provider instance
+    - Manually added for non-component External Service provider instances (e.g. Oracle DB) that do not know about k8s (rather than adding Endpoints)
+  - Different from Endpoints
+    - Singular, instead of plural (use -List, namespacing, and labels for grouping)
+    - Includes both Ready & NotReady endpoints
+    - Supports multiple and future conditions (in an extensible Conditions array)
+2. Modify ServiceSpec.Selector to select from both Pods and ExternalServiceProviders
+3. Add a new ExternalServiceProvider Controller that handles probing ExternalServiceProviders
+  - Probe results update ExternalServiceProvider.Status.Conditions
+4. Modify Endpoints Controller to manage ExternalServiceProvider Endpoints
+  - Readiness from ExternalServiceProvider.Status.Conditions analogous to PodStatus.Conditions
+5. Modify the Primary Components to self-register ExternalServiceProviders (one per component instance) instead of an Endpoint
+  - Group ExternalServiceProviders (component instances) with labels (e.g. "type=component", "component-type=core")
+6. Modify the Addon Components' Pod and Service registration to use similar labels (e.g. "type=component", "component-type=addon")
+7. Register component Services (with ServiceSpec.Selector) as part of cluster deployment
+  - Services of new components can also be added at runtime, before any instances exist. Each instance would self-register a ExternalServiceProviders.
+8. Deprecate `/componentsstatuses` in favor of requesting a filtered list of ExternalServiceProviders
+  - ExternalServiceProviders are external-only, but have generic conditions (liveness and readiness)
+  - Endpoints are external + internal, but only have readiness (for now)
+
+
+#### API Types
+
+The following is a draft of how the API models might change.
+
+```
+type ExternalServiceProvider struct {
+	TypeMeta
+	ObjectMeta
+	Spec ExternalServiceProviderSpec
+	Status ExternalServiceProviderStatus
+}
+
+type ExternalServiceProviderSpec struct {
+	Host string
+	Ports []ExternalServiceProviderPort
+	LivenessProbe *Probe
+	ReadinessProbe *Probe
+}
+
+type ExternalServiceProviderPort struct {
+	Name string // optional - for mapping from ServicePort
+	Protocol Protocol // TCP, UDP
+	Port int
+}
+
+type ExternalServiceProviderStatus struct {
+	Conditions []ExternalServiceProviderCondition
+}
+
+type ExternalServiceProviderCondition struct {
+	Type ExternalServiceProviderConditionType // Readiness, Liveness
+	Status ConditionStatus // True, False, Unknown
+	LastUpdateTime util.Time
+	LastTransitionTime util.Time
+	Reason string
+	Message string
+}
+
+type ExternalServiceProviderList struct {
+	TypeMeta
+	ListMeta
+	Items []ExternalServiceProvider
+}
+```
+
+```
+type ServiceStatus struct {
+	...
+	Conditions []ServiceCondition
+}
+
+type ServiceCondition struct {
```

Comment:
Yeah, there's several condition types and they don't all have the same fields either.